04-Protocol-Security.md

```diff
@@ -321,9 +321,11 @@ uint pt_len_to_ct_len(uint pt_len) {
 }
 ```
 
+- Note: The `message_length` (payload_length) in the encrypted Stratum message header always reflects the plaintext payload size. The size of the encrypted payload is implicitly understood to be message_length + MAC size for each block. This simplifies the decryption process and ensures clarity in interpreting frame data.
+
 #### Decrypting stratum message
 1. read exactly 22 bytes and decrypt into stratum frame or fail
-2. read `frame.message_length` number of bytes and decrypt into plaintext payload or fail
+2. read `frame.message_length` number of bytes (as specified in the decrypted header) and decrypt them into plaintext payload. If decryption fails, the process stops.
```

Comment:
The value `frame.message_length` should be first converted to the ciphertext length. That amount of data should be read.

Also, please move the code snippet from line 308 to this subsection.

Reply:
Hi @jakubtrnka if the frame.message length is converted to the ciphertext, is that cipher text then decrypted to a plaintext payload?

Reply:
I am trying to understand that section because I believed the frame.message_length to be encrypted and then decrypted on the fly to a plaintext payload

Reply:
The problem is that encrypting the data expands its length, because each 64kB chunk is MACed. So the question was what value should the msg header field refer to.

In my opinion, and opinion of others, the encryption should be independent and transparent operation to the message serialization.
For instance, a plain serialized message (plaintext payload) is 100 bytes long. its frame header should contain message_length == 100.
After the frame has been encrypted, its ciphertext payload is actually 116 bytes long, while the encrypted header is just en encryption of the plaintext header (with value of 100).
If the plaintext payload is 100000 bytes long, then its ciphertext is 100032 bytes long, etc...

Hence, upon the header decryption, the decryptor needs to calculate how many ciphertext data it should read in order to decrypt it into a `header.payload_length` long plaintext payload, which then would be deserialized into a stratum message.

Reply:
Alright. I understand now. Thank you very much, I have made the needed changes , please review @jakubtrnka